### ISSUE TITLE
ThresholdOptimizer returns different group results when given multiple sensitive features.

### ISSUE THREAD

**Issue by LeJit** [fairlearn/fairlearn#728]
#### Describe the bug

The ThresholdOptimizer returns incorrect group performance results when multiple sensitive features are given. If the multiple sensitive feature columns are concatenated into a single column, the results appear more inline with what we intuitively expect.
This bug was found when optimizing for `balanced_accuracy_score` with the fairness constraint of `equalized_odds`.
#### Steps/Code to Reproduce

A script to reproduce the error can be found on the [manandi/scipy_case_study](https://github.com/LeJit/fairlearn/blob/manandi/scipy_case_study/tutorial/scipy/case_study.py) branch.

To reproduce the error, change the `sensitive` variable between the two values specified in lines 23 and 24. 

#### Expected Results

<img width="542" alt="Postprocess_bug" src="https://user-images.githubusercontent.com/1124464/113729775-51cc4f80-96c5-11eb-803f-468249e493e5.png">

When the sensitive features are concatenated together, the `ThresholdOptimizer` yields results that mitigate disparity across sensitive attributes that align with intuition.

#### Actual Results

<img width="542" alt="Postprocess_separate" src="https://user-images.githubusercontent.com/1124464/113729777-51cc4f80-96c5-11eb-8e78-c56a9ab8803c.png">

Within racial groups, we see the ThresholdOptimizer achieves a false positive rate of 0 and false negative rate of 1 for multiple demographic groups. This result is unintuitive and leads to large between-group disparities.

#### Versions

OS: Windows
Python Version: 3.8.5
Fairlearn version: 0.6.0

**Comment by michaelamoako** [comment]
I had a similar issue with other combinations of optimization metric and constraints: 

I think because the constraints are related to parity (and not upperbounds) these types of solutions are **possible**. Commenting to track if progress is made on this, as I have the same question

**Comment by romanlutz** [comment]
I've managed to trace this to `_merge_columns` in `_input_validation.py`. Evidently, if you use `np.apply_along_axis` to merge multiple text labels it cuts them off. Example:
```
>>> np.apply_along_axis(lambda row: ",".join([str(i) for i in row]), axis=1, arr=np.array([["abc", "def"], ["abdddddddddddddddddddddddddddddddd", "danfaenofeinfaoieraj"], ["a", "b"]]))
array(['abc,def', 'abddddd', 'a,b'], dtype='<U7')
```
That's also what's happening in your example since some of the sensitive feature group names are fairly long. Clearly, that shouldn't be an issue, so I'm currently looking into what can be done. It seems like `np.apply_along_axis` may just not be terribly suitable for string operations. I'll dig a little deeper to figure out how to do this properly. Stay tuned!

**Comment by romanlutz** [comment]
I have a fix and I'm sending out a PR momentarily.

**Comment by romanlutz** [comment]
Closing this since #763 fixed the issue and v0.6.2 contains the fix.
